ruby/ql/lib/codeql/ruby/dataflow/FlowSummary.qll

```diff
@@ -158,3 +160,66 @@ abstract class SimpleSummarizedCallable extends SummarizedCallable {
 }
 
 class RequiredSummaryComponentStack = Impl::Public::RequiredSummaryComponentStack;
+
+/**
+ * Provides a set of special flow summaries to ensure that callbacks passed into
+ * library methods will be passed as `self` arguments into themeselves. That is,
+ * we are assuming that callbacks passed into library methods will be called, which is
+ * needed for flow through captured variables.
+ */
+private module LibraryCallbackSummaries {
+  private predicate libraryCall(CfgNodes::ExprNodes::CallCfgNode call) {
+    not exists(getTarget(call))
+  }
+
+  private class LibraryBlockMethod extends SummarizedCallable {
+    LibraryBlockMethod() { this = "<library method accepting a block>" }
+
+    final override MethodCall getACall() {
+      libraryCall(result.getAControlFlowNode()) and
```

Comment:
FYI I ran into an issue in JS where tricks like this would negatively affect the field-flow branch limit, since some calls then appear to have too many callees. Basically we're trying to model the behaviour of a single callable by treating it as two different call targets, each with different flows, but this is not equivalent to having a single callable with the union of those flows.

Since the default field-flow branch limit is 2, having 2 callees should be fine, and I'm guessing perhaps that's why this issue hasn't come up in the Ruby implementation yet. But if we did this trick one more time, it seems to me that things would stop working (assuming the callee also has >2 callers).

On the topic of the field-flow branch limit, the `InlineFlowTest.qll` file overrides it to 1000, which would hide this issue from ourselves. When I manually lower it to 2, some of the capture-related tests start failing with missing flows.

I'm not sure if the underlying issue needs to be fixed in this PR, but it would be great if we could remove this footgun from the test suite ASAP. Either by lowering the field-flow branch limit for capture-related tests, or add a query predicate to the test that reports all the places where we _only_ find a result with the higher branch limit.

Reply:
That's a good point. The same actually also applies for flow out (through): If we generate only a single unified callable, then it may have many return sites, which in turn means that the branch limit can kick in there as well. To work around this, we would have to synthesize a callable for each call, but I think that is difficult to do properly without running into negative recursion.

Reply:
As discussed offline, what I wrote above is actually not an issue after all.